### PR TITLE
feat: Azure DevOps logging commands

### LINKED ISSUE #38: Option to log warnings and errors with Azure DevOps logging commands
## Is your feature request related to a problem? Please describe

Often the package deployment task in the release pipeline will show errors and warnings when opened. However, Azure DevOps does not pick these up as warnings or errors on the pipeline itself.

## Describe the solution you'd like

The option to log errors and warnings with [Azure DevOps logging commands](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands)

### PR DESCRIPTION
**Pull request by ewingjm**
## Purpose
Resolves #38. Despite logging errors or warnings, Azure DevOps tasks deploying a package using this template would not show as failed or partially succeeded. In addition, warnings or errors were not visible on the pipelines themselves and were often hidden in the very large log generated by the template.

## Approach
If the deploy is running on Azure DevOps, it formats warnings and errors with [LogIssue](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logissue-log-an-error-or-warning) logging commands so that they're visible on the pipeline. In addition, if any errors are logged during the deployment, it will mark the task as failed. If no errors have occurred but there have been warnings, it will mark the task as partially succeeded.

## TODOs
- [x] Automated test coverage for new code
- [x] Documentation updated (if required)
- [x] Build and tests successful

### CODE REVIEW

File: src/Capgemini.PowerApps.PackageDeployerTemplate/PackageTemplateBase.cs [hunk 1]
```diff
@@ -448,12 +462,24 @@ private void ExecuteLifecycleEvent(string eventName, Action eventAction)
...
+            this.TraceLoggerAdapter.LogInformation($"{nameof(PackageTemplateBase)}.{lifecycleEvent} running...");
         }
 
         private void LogLifecycleEventEnd(string lifecycleEvent)
         {
-            this.PackageLog.Log($"{nameof(PackageTemplateBase)}.{lifecycleEvent} completed.", TraceEventType.Information);
+            this.TraceLoggerAdapter.LogInformation($"{nameof(PackageTemplateBase)}.{lifecycleEvent} completed.");
+        }
+
+        private void LogTaskCompleteResult()
```
Comment (tdashworth) [inline src/Capgemini.PowerApps.PackageDeployerTemplate/PackageTemplateBase.cs:473]: Will this double log? e.g.

`##[task.logissue type=error]##vso[task.complete result=Failed;]DONE`

Comment (ewingjm) [reply src/Capgemini.PowerApps.PackageDeployerTemplate/PackageTemplateBase.cs:473]: yup, good spot!

**Comment by tdashworth** [comment]
Could we add some unit tests for `AzureDevOpsTraceLoggerAdapter` to check for each log level?

**Review by markcunninghamuk — DISMISSED** [review]
Useful to see the TF_BUILD can be used, Not sure why I never saw it before on the predefined variables list. Good spot @tdashworth on the double log.

**Comment by ewingjm** [comment]
> Could we add some unit tests for `AzureDevOpsTraceLoggerAdapter` to check for each log level?

As agreed we won't do this as `TraceLogger` can't be mocked and to instantiate without mocking in the unit test appears to be very difficult. It requires a reference to `Microsoft.VisualBasic` and it seems the tests wouldn't be able to run in parallel as it can only write to a single file-based listener.

**Review by tdashworth — APPROVED** [review]

**Pull request merged**